### ISSUE TITLE
Apparence turns light (for an instant) to when back and forth on iOS with SDK39

### ISSUE THREAD

**Issue by outaTiME** [expo/expo#10815]
Hi guys, 

For some reason the `useColorScheme` on iOS returns "light" (for a moment) when the application goes to background and returns to foreground, I suppose that this is something that the SDK39 introduces because I had no problem with the previous version.

It could be that it is related to a facebook issue:

https://github.com/facebook/react-native/issues/28525

here I leave an video of the problem:

![RPReplay_Final1603742498](https://user-images.githubusercontent.com/140472/97225008-c7e09800-17b0-11eb-95d4-c2e10f63c1ca.gif)

Alternatively I tried with the native `useColorTheme` and had the same unwanted effect.

**Comment by outaTiME** [comment]
Related with: https://github.com/expo/react-native-appearance/issues/53

**Comment by brentvatne** [comment]
I'm not sure how https://github.com/expo/react-native-appearance/issues/53 is supposed to be related. This is an issue with how react-native upstream afaict. If you use react-native-appearance instead, this shouldn't be a problem because of this code: https://github.com/expo/react-native-appearance/blob/9dcd12390e43f1148a4901c70daf1a9085b5e5c1/ios/Appearance/RNCAppearanceProvider.m#L27-L35

**Comment by brentvatne** [comment]
copying over from the other issue referenced above, @outaTiME wrote:

<hr />

Sure, using this minimal example, with `"userInterfaceStyle": "automatic"` on app.json the logs shows the `colorScheme` goes to light and then go back dark (initial).

```js
import * as React from 'react';
import { StyleSheet, View, Text, Button } from 'react-native';
import { Appearance, AppearanceProvider, useColorScheme } from 'react-native-appearance';
import { StatusBar } from 'expo-status-bar';

export default () => {
  const colorScheme = useColorScheme();
  console.log('>>> colorScheme', colorScheme);
  const isDark = colorScheme === 'dark';
  return (
    <AppearanceProvider>
      <StatusBar style="auto" animated />
      <View style={{ flex: 1, backgroundColor: isDark ? '#333' : '#f1f1f1' }} />
    </AppearanceProvider>
  );
};
```

```
>>> colorScheme dark
>>> colorScheme light
>>> colorScheme dark
>>> colorScheme dark
>>> colorScheme light
>>> colorScheme dark
```

in the log above each of the light entries means that the application goes into background, and goes back to dark when it returns to foreground.

in complex structures this possibly generates unnecessary redrawings (as is the case of the image reported in the issue: https://github.com/expo/expo/issues/10815)

**Comment by jasonaibrahim** [comment]
if anyone needs a workaround while this is sorted out, here it is as a hook that can be used in lieu of the rn `useColorScheme` hook. it throttles the color scheme change events so that the more recent one is used as the `colorScheme` value, since the first event seems to be incorrect but subsequent values are correct

```typescript jsx
import { Appearance, ColorSchemeName } from 'react-native';
import { useEffect, useRef, useState } from 'react';

export default function useColorScheme(delay = 500): NonNullable<ColorSchemeName> {
  const [colorScheme, setColorScheme] = useState(Appearance.getColorScheme());

  let timeout = useRef<NodeJS.Timeout | null>(null).current;

  useEffect(() => {
    Appearance.addChangeListener(onColorSchemeChange);

    return () => {
      resetCurrentTimeout();
      Appearance.removeChangeListener(onColorSchemeChange);
    };
  }, []);

  function onColorSchemeChange(preferences: Appearance.AppearancePreferences) {
    resetCurrentTimeout();

    timeout = setTimeout(() => {
      setColorScheme(preferences.colorScheme);
    }, delay);
  }

  function resetCurrentTimeout() {
    if (timeout) {
      clearTimeout(timeout);
    }
  }

  return colorScheme as NonNullable<ColorSchemeName>;
}

```

**Comment by outaTiME** [comment]
great @jasonaibrahim, 
i made few tweak usign lodash and better memoization to reduce size, at this moment the only way i know to "workaround" this issue is throttled the `Appearance` updates 😞 

```js
useColorScheme(delay = 250) {
  const [colorScheme, setColorScheme] = React.useState(
    Appearance.getColorScheme()
  );
  const onColorSchemeChange = React.useCallback(
    _.throttle(
      ({ colorScheme }) => {
        setColorScheme(colorScheme);
      },
      delay,
      {
        leading: false,
        // trailing: true,
      }
    ),
    []
  );
  React.useEffect(() => {
    Appearance.addChangeListener(onColorSchemeChange);
    return () => {
      onColorSchemeChange.cancel();
      Appearance.removeChangeListener(onColorSchemeChange);
    };
  }, []);
  return colorScheme;
}
```

BW, 250ms on throttle invocations works for me

**Comment by brentvatne** [comment]
i think that is a reasonable solution for now. keep in mind that this is to an extent actually the intended behavior of the ios api - if i remember correctly, the reason it switches between light and dark mode is to take a screenshot of your app in both states so that it can show the appropriate version in the multitasker if between color schemes on the os and then look in the multitasker.

**Comment by outaTiME** [comment]
@brentvatne disable the behavior at react-native level or react-native-appearance package?

**Comment by brentvatne** [comment]
@outaTiME - i don't understand the question

**Comment by outaTiME** [comment]
@brentvatne sorry, previously you said that in your opinion it was best to disable this behavior, what did you mean? using the workaround above or disable it to another level?

**Comment by brentvatne** [comment]
yeah using the workaround mentioned above. thanks for sharing that!

**Comment by outaTiME** [comment]
@brentvatne lately even though this workaround very rarely the appearance change continues to occur (even with higher throttle invocations), imho there should be a solution at the react-native level or not?

**Comment by brentvatne** [comment]
yes i think there should be. maybe you can post a react-native issue about it?

**Comment by outaTiME** [comment]
There seems to be some reports already in the react-native repository:

https://github.com/facebook/react-native/issues/28525
https://github.com/facebook/react-native/issues/28823

but nothing formally solved yet 😢

**Comment by Aryk** [comment]
Yeah, I've been experiencing this too, and yes I could implement the throttling (I'm seeing 250 and 600) as good numbers up there...but using a timer is a pretty ugly hack. I hope this gets resolved soon. And at the very least if it's in the API spec, then should come with another variable like "forScreenshot" so we know it doesn't reflect the actual device settings.

So for me, it happens like 3% of the time...is there anyway to make this consistently happen so that I can test my fix?

**Comment by outaTiME** [comment]
hi @Aryk, 
I suppose this is something that the os handles, I don't think you could consistently test it, even if the device is on battery saving, this behavior can be affected.

Do you have a better workaround than was proposed? I think we depend on the goodwill of the react-native developers to fix this problem 😢

**Comment by Aryk** [comment]
Got it, it's been driving me nuts. 😆 

Your solution is exactly what I was thinking about doing...for some reason though...I havent had a lot of luck with the throttle-debounce from lodash so I switched to   "throttle-debounce": "^2.1.0". So if you find doesn't quite work, try that library.

Also not a big deal..but just fyi you can do:

```ts
const subscription = Appearance.addChangeListener...
return () => subscription.remove(); 
```

in your useEffect

**Comment by outaTiME** [comment]
I really don't like this workaround too much, on the other hand the snippet above uses the `Appearance` from `react-native` and you cannot do a `.remove` as in `react-native-appearance` from Expo (be careful with that), 
thanks 🙏

**Comment by Aryk** [comment]
> @Aryk I really don't like this workaround too much, on the other hand the snippet above uses the `Appearance` from `react-native` and you cannot do a `.remove` as in `react-native-appearance` from Expo (be careful with that),
> thanks 🙏

Ah didn't know that. yeah I'm using react-native-appearance bc the new version fixes some issues on Android.

**Comment by outaTiME** [comment]
indeed, internally I am using react-native-appearance as dependency 💪

**Comment by badbod99** [comment]
Running a brand new Expo project using Expo init, with the sample from react-native-appearance, useColorScheme **always** returns light on iPhone 12 set to Dark theme.

**Comment by brentvatne** [comment]
@badbod99 - you need to set `userInterfaceStyle` in app.json, it's in the readme. https://github.com/expo/react-native-appearance#managed-expo-project

**Comment by Aryk** [comment]
Hey guys...Im on SDK 41 and it's still happening for me...I'm at 550ms with the solution here:

https://github.com/expo/expo/issues/10815#issuecomment-719113200

But this is still happening, so gonna try 650ms...

**Comment by outaTiME** [comment]
This workaround working for me now:

https://github.com/facebook/react-native/issues/28525#issuecomment-841812810

Here complete snippet:

```
import { Appearance } from 'react-native';

useColorScheme() {
  const [colorScheme, setColorScheme] = React.useState(
    Appearance.getColorScheme()
  );
  const onColorSchemeChange = React.useCallback(() => {
    setColorScheme(Appearance.getColorScheme());
  }, []);
  React.useEffect(() => {
    Appearance.addChangeListener(onColorSchemeChange);
    return () => {
      Appearance.removeChangeListener(onColorSchemeChange);
    };
  }, []);
  return colorScheme;
}
```

**Comment by outaTiME** [comment]
> This workaround working for me now:
> 
> [facebook/react-native#28525 (comment)](https://github.com/facebook/react-native/issues/28525#issuecomment-841812810)
> 
> Here complete snippet:
> 
> ```js
> import { Appearance } from 'react-native';
> 
> useColorScheme() {
>   const [colorScheme, setColorScheme] = React.useState(
>     Appearance.getColorScheme()
>   );
>   const onColorSchemeChange = React.useCallback(() => {
>     setColorScheme(Appearance.getColorScheme());
>   }, []);
>   React.useEffect(() => {
>     Appearance.addChangeListener(onColorSchemeChange);
>     return () => {
>       Appearance.removeChangeListener(onColorSchemeChange);
>     };
>   }, []);
>   return colorScheme;
> }
> ```

Forget this, after a day of testing, very randomly it keeps happening, 

I had to go back to the original workaround 😭

**Comment by Aryk** [comment]
@outaTiME what is your "original workaround". Is it this one: https://github.com/expo/expo/issues/10815#issuecomment-719113200

Btw my delay on my throttle is now 650 and seems to still occasionally happen

**Comment by kickbk** [comment]
Seeing similar issues with Expo Go and a published app, despite using the suggested workaround. Sometimes switching back to the app has appearance stuck in dark mode for react navigation, while the other parts of the app are in light mode. This should probably be improved so it works as other native apps do without flashing or workarounds.

> the intended behavior of the ios api - if i remember correctly, the reason it switches between light and dark mode is to take a screenshot of your app in both states so that it can show the appropriate version in the multitasker if between color schemes on the os and then look in the multitasker.

That may be the case, but I switch in and out of other apps that support dark mode, like whatsapp, and there's never any flashing.

**Comment by umrashrf** [comment]
On the contrary, iPhone simulator always has light and iPhone device has always dark for me when I useColorScheme.

Please see this https://snack.expo.dev/@umrashrf89/example This returns always light on Expo client on iPhone.

**Comment by TomCorvus** [comment]
@jasonaibrahim solution works for me. I just change his code to avoid ESLint warnings:
```
import { Appearance, ColorSchemeName } from 'react-native';
import { useCallback, useEffect, useRef, useState } from 'react';

export const useColorScheme = (delay = 500): NonNullable<ColorSchemeName> => {
	const [colorScheme, setColorScheme] = useState(Appearance.getColorScheme());
	const timeout: any = useRef<NodeJS.Timeout | null>(null);

	const resetCurrentTimeout = useCallback(() => {
		if (timeout?.current) {
			clearTimeout(timeout.current);
		}
	}, [timeout]);

	const onColorSchemeChange = useCallback(
		(preferences: Appearance.AppearancePreferences) => {
			resetCurrentTimeout();

			timeout.current = setTimeout(() => {
				setColorScheme(preferences.colorScheme);
			}, delay);
		},
		[delay, resetCurrentTimeout]
	);

	useEffect(() => {
		const AppearanceChange = Appearance.addChangeListener(onColorSchemeChange);
		return () => {
			resetCurrentTimeout();
			if (AppearanceChange) AppearanceChange.remove();
		};
	}, [onColorSchemeChange, resetCurrentTimeout]);

	return colorScheme as NonNullable<ColorSchemeName>;
};

export default useColorScheme;
```

**Comment by IngyuTae** [comment]
Shorter solution

```ts
import { useEffect, useState } from 'react'
import {
  ColorSchemeName,
  useColorScheme as _useColorScheme,
} from 'react-native'

// The useColorScheme value is always either light or dark, but the built-in
// type suggests that it can be null. This will not happen in practice, so this
// makes it a bit easier to work with.
export default function useColorScheme(): NonNullable<ColorSchemeName> {
  const _colorScheme = _useColorScheme()
  const [colorScheme, setColorScheme] = useState(_colorScheme)

  useEffect(() => {
    const timer = setTimeout(() => setColorScheme(_colorScheme), 100)
    return () => clearTimeout(timer)
  }, [_colorScheme])

  return colorScheme as NonNullable<ColorSchemeName>
}
```

**Comment by efstathiosntonas** [comment]
Do not close the issue, still a bug

**Comment by matusca96** [comment]
Just in case anyone else is having the app reseted, this could be the issue, to fix it just replace the `useColorScheme` from `react-native`to the custom one from this [comment](https://github.com/expo/expo/issues/10815#issuecomment-719113200)

**Comment by sreuter** [comment]
It seems like the exact bug is back in SDK 49. Anyone else seeing this?

**Comment by Jkitche** [comment]
I've also been fighting with this on SDK 49

**Comment by sreuter** [comment]
@brentvatne As this broke with Expo SDK 49 again, is the recommended way still switching to `react-native-appearance`? This appeared to be fixed at some point but just popped up with the latest release, hence I'm asking.

**Comment by brentvatne** [comment]
some related issues on the react-native repo:
- https://github.com/facebook/react-native/issues/35972
- https://github.com/facebook/react-native/issues/36713

@alanjhughes opened a pr for one of these (https://github.com/facebook/react-native/pull/38214) but i'm not sure if it'll fix the other. he will follow up on this

**Comment by brentvatne** [comment]
@sreuter - if you copy [this diff](https://github.com/facebook/react-native/pull/38214/files) over to your node_modules and use patch-package, can you let us know if this resolves your issue?
